File: android/app/src/main/java/com/clingsync/android/presentation/MainReducer.kt
package com.clingsync.android.presentation

import com.clingsync.android.FileStatus
import com.clingsync.android.UploadInfo
import com.clingsync.android.validateHostUrl

// Pure transition table for user intents and internal completions. Returns the
// next state plus the side effects the ViewModel should perform. No IO, no
// Android framework, no coroutines: every rule is one assertable function.
object MainReducer {
    fun reduce(
        state: MainUiState,
        event: MainEvent,
    ): Reduction =
        when (event) {
            is MainEvent.FileSelectionChanged ->
                only(
                    state.copy(
                        selectedPaths =
                            if (event.selected) {
                                state.selectedPaths + event.path
                            } else {
                                state.selectedPaths - event.path
                            },
                    ),
                )

            MainEvent.SelectAllClicked -> only(state.copy(selectedPaths = state.selectAllTargets))

            is MainEvent.SearchQueryChanged ->
                only(state.copy(searchQuery = event.query, selectedPaths = emptySet()))

            MainEvent.SearchToggled -> {
                val opening = !state.showSearch
                only(
                    state.copy(
                        showSearch = opening,
                        searchQuery = if (opening) state.searchQuery else "",
                        selectedPaths = if (opening) emptySet() else state.selectedPaths,
                    ),
                )
            }

            MainEvent.SearchCleared -> only(state.copy(searchQuery = ""))

            MainEvent.RefreshClicked -> Reduction(state, listOf(Effect.LoadFiles))

            MainEvent.UploadClicked -> upload(state)

            MainEvent.AbortClicked ->
                Reduction(state.copy(isUploadInitiated = false), listOf(Effect.CancelUpload))

            is MainEvent.PermissionResult ->
                if (event.granted) {
                    Reduction(state.copy(hasPermission = true), listOf(Effect.LoadFiles))
                } else {
                    only(state.copy(hasPermission = false))
                }

            MainEvent.LoadingStarted -> only(state.copy(isLoadingFiles = true))

            is MainEvent.FilesLoaded ->
                only(
                    state.copy(
                        isLoadingFiles = false,
                        files = event.files,
                        overlay =
                            if (event.needsStoragePermission && state.overlay is Overlay.None) {
                                Overlay.StoragePermission
                            } else {
                                state.overlay
                            },
                    ),
                )

            is MainEvent.ScanStarted ->
                only(
                    state.copy(
                        isScanning = true,
                        scanProgress = 0 to event.paths.size,
                        fileStatus = state.fileStatus + event.paths.associateWith { FileStatus.Scanning },
                    ),
                )

            is MainEvent.ScanProgress ->
                only(
                    state.copy(
                        fileStatus = state.fileStatus + event.statuses,
                        scanProgress = event.processed to event.total,
                    ),
                )

            is MainEvent.ScanCompleted ->
                only(
                    state.copy(
                        fileStatus = state.fileStatus + event.statuses,
                        scanProgress = null,
                        isScanning = false,
                    ),
                )

            is MainEvent.ScanFailed -> scanFailed(state, event)

            MainEvent.SettingsClicked -> only(state.copy(showSettings = true))

            MainEvent.SettingsDismissed -> only(state.copy(showSettings = false))

            is MainEvent.SettingsSaved -> settingsSaved(state, event)

            is MainEvent.SettingsTestConnection -> {
                val urlError = validateHostUrl(event.settings.hostUrl)
                if (urlError != null) {
                    only(state.copy(overlay = Overlay.Error("Invalid Host URL", urlError)))
                } else {
                    Reduction(state.copy(settings = event.settings), listOf(Effect.Connect))
                }
            }

            MainEvent.ConnectClicked -> Reduction(state, listOf(Effect.Connect))

            MainEvent.ConnectStarted -> only(state.copy(isConnecting = true))

            MainEvent.ConnectSucceeded ->
                only(state.copy(isConnecting = false, isConnected = true))

            is MainEvent.ConnectFailed ->
                only(
                    state.copy(
                        isConnecting = false,
                        isConnected = false,
                        overlay =
                            if (state.overlay is Overlay.None) {
                                Overlay.Error("Connection Error", event.message)
                            } else {
                                state.overlay
                            },
                    ),
                )

            is MainEvent.ShowPassphrasePrompt ->
                only(state.copy(overlay = Overlay.Passphrase(event.showKeychainOption)))

            is MainEvent.PassphraseEntered -> only(state.copy(overlay = Overlay.None))

            MainEvent.PassphraseDismissed -> only(state.copy(overlay = Overlay.None))

            // Biometric results are consumed by the ViewModel (to drive the open
            // flow); they carry no state change of their own.
            is MainEvent.PassphraseLoaded -> only(state)

            is MainEvent.PassphraseLoadFailed -> only(state)

            MainEvent.ShowS3Prompt -> only(state.copy(overlay = Overlay.S3Credentials))

            is MainEvent.S3CredentialsEntered -> only(state.copy(overlay = Overlay.None))

            MainEvent.S3CredentialsDismissed -> only(state.copy(overlay = Overlay.None))

            MainEvent.ErrorDismissed -> only(state.copy(overlay = Overlay.None))

            MainEvent.StoragePermissionDismissed -> only(state.copy(overlay = Overlay.None))

            MainEvent.OpenStorageSettingsClicked ->
                Reduction(state.copy(overlay = Overlay.None), listOf(Effect.OpenStorageSettings))
        }

    private fun only(state: MainUiState): Reduction = Reduction(state)

    private fun upload(state: MainUiState): Reduction {
        val paths = state.selectedFiles.map { it.absolutePath }
        if (paths.isEmpty()) return only(state)
        return Reduction(
            state.copy(
                isUploadInitiated = true,
                scanProgress = null,
                uploadInfo = UploadInfo(currentFile = null, currentIndex = 0, totalFiles = paths.size),
                currentUploadPaths = paths.toSet(),
                fileStatus = state.fileStatus + paths.associateWith { FileStatus.Waiting },
                selectedPaths = emptySet(),
            ),
            listOf(Effect.EnqueueUpload(paths, state.settings.author)),
        )
    }

    private fun scanFailed(
        state: MainUiState,
        event: MainEvent.ScanFailed,
    ): Reduction {
        val reverted =
            event.paths.filter { state.fileStatus[it] is FileStatus.Scanning }.associateWith { FileStatus.New }
        return only(
            state.copy(
                fileStatus = state.fileStatus + reverted,
                scanProgress = null,
                isScanning = false,
                overlay =
                    if (state.overlay is Overlay.None) {
                        Overlay.Error("File Scanning Error", "Some files could not be scanned: ${event.message}")
                    } else {
                        state.overlay
                    },
            ),
        )
    }

    private fun settingsSaved(
        state: MainUiState,
        event: MainEvent.SettingsSaved,
    ): Reduction {
        val urlError = validateHostUrl(event.settings.hostUrl)
        if (urlError != null) {
            return only(state.copy(overlay = Overlay.Error("Invalid Host URL", urlError)))
        }
        val oldId = state.settings.repositoryID()
        val repositoryChanged = oldId != event.settings.repositoryID()
        val sourceChanged =
            state.settings.sourceDirectory != event.settings.sourceDirectory ||
                state.settings.mediaOnly != event.settings.mediaOnly
        val base = state.copy(settings = event.settings, showSettings = false)
        val persist = Effect.PersistSettings(event.settings)
        return when {
            repositoryChanged ->
                Reduction(
                    base.copy(fileStatus = emptyMap(), isConnected = false, selectedPaths = emptySet()),
                    listOf(persist, Effect.InvalidateRepository(oldId), Effect.LoadFiles),
                )
            sourceChanged ->
                Reduction(
                    base.copy(fileStatus = emptyMap(), selectedPaths = emptySet()),
                    listOf(persist, Effect.LoadFiles),
                )
            else -> Reduction(base, listOf(persist))
        }
    }
}

File: ios/ClingSync/MainReducer.swift
import Foundation

// Pure transition table for user intents and internal completions. Returns the
// next state plus the side effects the store should perform. No IO, no Photos,
// no bridge, no clock: every rule is one assertable function.
enum MainReducer {
    // swiftlint:disable:next cyclomatic_complexity function_body_length
    static func reduce(_ state: AppState, _ event: MainEvent) -> Reduction {
        switch event {
        case .fileSelectionChanged(let id, let selected):
            var next = state
            if selected {
                next.selectedIds.insert(id)
            } else {
                next.selectedIds.remove(id)
            }
            return .only(next)

        case .selectAllClicked:
            var next = state
            next.selectedIds = state.selectAllTargets
            return .only(next)

        case .deselectAllClicked:
            var next = state
            next.selectedIds = []
            return .only(next)

        case .searchQueryChanged(let query):
            var next = state
            next.searchQuery = query
            next.selectedIds = []
            return .only(next)

        case .searchToggled:
            let opening = !state.showSearch
            var next = state
            next.showSearch = opening
            next.searchQuery = opening ? state.searchQuery : ""
            next.selectedIds = opening ? [] : state.selectedIds
            return .only(next)

        case .searchCleared:
            var next = state
            next.searchQuery = ""
            return .only(next)

        case .refreshClicked:
            return Reduction(state: state, effects: [.loadFiles])

        case .uploadClicked:
            return upload(state)

        case .abortClicked:
            var next = state
            next.isUploadInitiated = false
            return Reduction(state: next, effects: [.cancelUpload])

        case .loadingStarted:
            var next = state
            next.isLoadingFiles = true
            return .only(next)

        case .filesLoaded(let files):
            var next = state
            next.isLoadingFiles = false
            next.files = files
            return .only(next)

        case .scanStarted(let ids):
            var next = state
            next.isScanning = true
            next.scanProgress = ScanProgress(processed: 0, total: ids.count)
            for id in ids {
                next.fileStatus[id] = .checking
            }
            return .only(next)

        case .scanProgress(let processed, let total, let statuses):
            var next = state
            next.fileStatus.merge(statuses) { _, new in new }
            next.scanProgress = ScanProgress(processed: processed, total: total)
            return .only(next)

        case .scanCompleted(let statuses):
            var next = state
            next.fileStatus.merge(statuses) { _, new in new }
            next.scanProgress = nil
            next.isScanning = false
            return .only(next)

        case .scanFailed(let message, let ids):
            return scanFailed(state, message: message, ids: ids)

        case .settingsClicked:
            var next = state
            next.showSettings = true
            return .only(next)

        case .settingsDismissed:
            var next = state
            next.showSettings = false
            return .only(next)

        case .settingsSaved(let configuration):
            return settingsSaved(state, configuration: configuration)

        case .connectClicked:
            return Reduction(state: state, effects: [.connect])

        case .connectStarted:
            var next = state
            next.isConnecting = true
            return .only(next)

        case .connectSucceeded:
            var next = state
            next.isConnecting = false
            next.isConnected = true
            next.phase = .ready
            return .only(next)

        case .connectFailed(let message):
            var next = state
            next.isConnecting = false
            next.isConnected = false
            next.phase = .connectionFailed(message)
            return .only(next)

        case .errorDismissed:
            var next = state
            next.overlay = .none
            return .only(next)
        }
    }

    private static func upload(_ state: AppState) -> Reduction {
        let ids = state.selectedFiles.map(\.id)
        if ids.isEmpty {
            return .only(state)
        }
        var next = state
        next.isUploadInitiated = true
        next.scanProgress = nil
        next.uploadInfo = UploadInfo(currentFile: nil, currentIndex: 0, totalFiles: ids.count)
        next.currentUploadIds = Set(ids)
        for id in ids {
            next.fileStatus[id] = .waiting
        }
        next.selectedIds = []
        return Reduction(state: next, effects: [.enqueueUpload(ids: ids, author: state.configuration.author)])
    }

    private static func scanFailed(_ state: AppState, message: String, ids: [String]) -> Reduction {
        var next = state
        for id in ids where state.fileStatus[id] == .checking {
            next.fileStatus[id] = .new
        }
        next.scanProgress = nil
        next.isScanning = false
        if state.overlay == .none {
            next.overlay = .error(
                title: "File Scanning Error",
                message: "Some files could not be scanned: \(message)")
        }
        return .only(next)
    }

    private static func settingsSaved(_ state: AppState, configuration: RepositoryConfiguration) -> Reduction {
        if let urlError = validateHostURL(configuration.hostURL) {
            var next = state
            next.overlay = .error(title: "Invalid Host URL", message: urlError)
            return .only(next)
        }
        let oldId = state.configuration.repositoryID
        let repositoryChanged = oldId != configuration.repositoryID
        var base = state
        base.configuration = configuration
        base.showSettings = false
        // First-run: only a successful connect sets `.ready` otherwise, so saving
        // valid settings from the Welcome screen must move on by itself.
        if configuration.isConfigured, base.phase == .needsSettings {
            base.phase = .ready
        }
        let persist = Effect.persistSettings(configuration)
        if repositoryChanged {
            base.fileStatus = [:]
            base.isConnected = false
            base.selectedIds = []
            return Reduction(
                state: base,
                effects: [persist, .invalidateRepository(repositoryID: oldId), .loadFiles])
        }
        return Reduction(state: base, effects: [persist])
    }
}
